File: shared/src/main/java/ru/privatenull/pnauth/extension/AuthOperationContext.java
package ru.privatenull.pnauth.extension;
import java.util.Map;
import java.util.UUID;
public record AuthOperationContext(
        AuthOperation operation, AuthPhase phase, UUID uniqueId, String username, String ip, Map<String, String> attributes
) {
    public AuthOperationContext {
        attributes = attributes == null ? Map.of() : Map.copyOf(attributes);
    }
    public AuthOperationContext(AuthOperation operation, UUID uniqueId, String username, String ip, Map<String, String> attributes) {
        this(operation, AuthPhase.BEFORE_EXECUTION, uniqueId, username, ip, attributes);
    }
    public static AuthOperationContext user(AuthOperation operation, UUID id, String username, String ip) {
        return new AuthOperationContext(operation, AuthPhase.BEFORE_EXECUTION, id, username, ip, Map.of());
    }
    public AuthOperationContext at(AuthPhase value) { return new AuthOperationContext(operation, value, uniqueId, username, ip, attributes); }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/extension/AuthOperationContext.kt
package ru.privatenull.pnauth.extension

import java.util.UUID

@JvmRecord
data class AuthOperationContext @JvmOverloads constructor(
    val operation: AuthOperation,
    val phase: AuthPhase = AuthPhase.BEFORE_EXECUTION,
    val uniqueId: UUID?,
    val username: String?,
    val ip: String?,
    val attributes: Map<String, String> = emptyMap()
) {
    constructor(
        operation: AuthOperation,
        uniqueId: UUID?,
        username: String?,
        ip: String?,
        attributes: Map<String, String>?
    ) : this(operation, AuthPhase.BEFORE_EXECUTION, uniqueId, username, ip, attributes ?: emptyMap())

    fun at(value: AuthPhase): AuthOperationContext {
        return AuthOperationContext(operation, value, uniqueId, username, ip, attributes)
    }

    companion object {
        @JvmStatic
        fun user(operation: AuthOperation, id: UUID?, username: String?, ip: String?): AuthOperationContext {
            return AuthOperationContext(operation, AuthPhase.BEFORE_EXECUTION, id, username, ip, emptyMap())
        }
    }
}
